enricher/javareach/javareach.go

```diff
@@ -0,0 +1,358 @@
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package javareach
+
+import (
+	"archive/zip"
+	"bufio"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"io/fs"
+	"maps"
+	"os"
+	"path/filepath"
+	"slices"
+	"strings"
+
+	"github.com/google/osv-scalibr/enricher"
+	"github.com/google/osv-scalibr/extractor"
+	"github.com/google/osv-scalibr/extractor/filesystem/language/java/archive"
+	archivemeta "github.com/google/osv-scalibr/extractor/filesystem/language/java/archive/metadata"
+	"github.com/google/osv-scalibr/inventory"
+	"github.com/google/osv-scalibr/log"
+	"github.com/google/osv-scalibr/plugin"
+)
+
+const (
+	// Name is the unique name of this detector.
+	Name        = "javareach"
+	MetaDirPath = "META-INF"
+)
+
+var (
+	ManifestFilePath = filepath.Join(MetaDirPath, "MANIFEST.MF")
+	MavenDepDirPath  = filepath.Join(MetaDirPath, "maven")
+	ServiceDirPath   = filepath.Join(MetaDirPath, "services")
+
+	ErrMavenDependencyNotFound = errors.New(MavenDepDirPath + " directory not found")
+)
+
+type Enricher struct{}
+
+func (Enricher) Name() string {
+	return Name
+}
+
+func (Enricher) Version() int {
+	return 0
+}
+
+func (Enricher) Requirements() *plugin.Capabilities {
+	return &plugin.Capabilities{
+		Network:  plugin.NetworkOnline,
+		DirectFS: false,
+	}
+}
+
+func (Enricher) RequiredPlugins() []string {
+	return []string{archive.Name}
+}
+
+func (Enricher) Enrich(ctx context.Context, input *enricher.ScanInput, inv *inventory.Inventory) error {
+	jars := make(map[string]struct{})
+	for i := range inv.Packages {
+		if inv.Packages[i].Extractor.Name() == archive.Name {
+			jars[inv.Packages[i].Locations[0]] = struct{}{}
+		}
+	}
+
+	for jar := range jars {
+		err := enumerateReachabilityForJar(ctx, jar, input, inv)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func fmtJavaInventory(i *extractor.Package) string {
+	return fmt.Sprintf("%s:%s", i.Metadata.(*archivemeta.Metadata).GroupID, i.Name)
+}
+
+func enumerateReachabilityForJar(ctx context.Context, jarPath string, input *enricher.ScanInput, inv *inventory.Inventory) error {
+	var allDeps []*extractor.Package
+
+	for i := range inv.Packages {
+		if inv.Packages[i].Locations[0] == jarPath {
+			allDeps = append(allDeps, inv.Packages[i])
+		}
+	}
+
+	slices.SortFunc(allDeps, func(i1 *extractor.Package, i2 *extractor.Package) int {
+		return strings.Compare(fmtJavaInventory(i1), fmtJavaInventory(i2))
+	})
+	for _, dep := range allDeps {
+		log.Debug("extracted dep",
+			"group id", dep.Metadata.(*archivemeta.Metadata).GroupID, "artifact id", dep.Name, "version", dep.Version)
+	}
+
+	// Unpack .jar
+	tmpDir, err := os.MkdirTemp("", "")
```

Comment:
Name the temp dir osv-scalibr-javareach or something similar to make it easier to debug.